pkg/wakaama/contrib/lwm2m_client_objects.c

```diff
@@ -0,0 +1,111 @@
+ *              Wakaama from a LwM2M client.
+ *
+ * @author      Leandro Lanzieri <leandro.lanzieri@haw-hamburg.de>
+ * @}
+ */
+
+#include "kernel_defines.h"
+#include "net/lwm2m.h"
+
+#include "lwm2m_client.h"
+#include "lwm2m_client_objects.h"
+
+/* These functions are defined by the objects (object_security.c,
+ * object_server.c, and object_access_control.c are implemented by the Wakaama
+ * package. device.c can be found in 'contrib/objects') */
+lwm2m_object_t *get_security_object(int server_id, const char *server_uri,
+                                    char *bs_psk_id, char *psk,
+                                    uint16_t psk_len, bool is_bootstrap);
+lwm2m_object_t *get_server_object(int server_id, const char *binding,
+                                  int lifetime, bool storing);
+lwm2m_object_t *lwm2m_get_object_device(void);
+lwm2m_object_t *acc_ctrl_create_object(void);
+void acl_ctrl_free_object(lwm2m_object_t * objectP);
+bool acc_ctrl_obj_add_inst(lwm2m_object_t* accCtrlObjP, uint16_t instId,
+                           uint16_t acObjectId, uint16_t acObjInstId,
+                           uint16_t acOwner);
+bool acc_ctrl_oi_add_ac_val(lwm2m_object_t* accCtrlObjP, uint16_t instId,
+                            uint16_t acResId, uint16_t acValue);
+
+lwm2m_object_t *lwm2m_client_get_security_object(
+                        lwm2m_client_data_t *client_data)
+{
+    lwm2m_object_t *ret;
+    char *server_uri = LWM2M_SERVER_URI;
+    int server_id = LWM2M_SERVER_ID;
+    uint16_t psk_len = -1;
+    char *psk_buffer = NULL;
+    char *psk_id = NULL;
+
+    ret = get_security_object(server_id, server_uri, psk_id, psk_buffer,
+                              psk_len, IS_ACTIVE(LWM2M_BOOTSTRAP));
+
+    client_data->obj_security = ret;
+    return ret;
+}
+
+lwm2m_object_t *lwm2m_client_get_server_object(
+                        lwm2m_client_data_t *client_data)
+{
+    (void)client_data;
+    lwm2m_object_t *ret;
+    int server_id = LWM2M_SERVER_ID;
+    int lifetime = LWM2M_DEVICE_TTL;
+
+    ret = get_server_object(server_id, LWM2M_DEVICE_BINDINGS, lifetime, false);
+    return ret;
+}
+
+lwm2m_object_t *lwm2m_client_get_device_object(
+                        lwm2m_client_data_t *client_data)
+{
+    (void)client_data;
+    return lwm2m_get_object_device();
+}
+
+lwm2m_object_t *lwm2m_client_get_acc_ctrl_object(
+                        lwm2m_client_data_t *client_data)
+{
+    (void)client_data;
+    lwm2m_object_t *ret;
+    uint16_t instance_id = 0;
+
+    ret = acc_ctrl_create_object();
+
+    if (!ret) {
+        goto out;
+    }
+
+    /* add an instance of access control list for the 'device' object */
+    if (!acc_ctrl_obj_add_inst(ret, instance_id, 3, 0, LWM2M_SERVER_ID)) {
+        goto free_out;
+    }
+
+    const uint16_t bits = LWM2M_ACC_CTRL_READ | LWM2M_ACC_CTRL_WRITE |
+                          LWM2M_ACC_CTRL_EXECUTE;
+    /* give Read, Write, Execute and Delete access */
```

Comment:
Did you intend to give delete access?

Reply:
So, I looked this up on the [LwM2M 1.0 specification](http://www.openmobilealliance.org/release/LightweightM2M/V1_0-20170208-A/OMA-TS-LightweightM2M-V1_0-20170208-A.pdf) (section 7.3):
> "In the particular case where a single LwM2M Server Account exists in the LwM2M Client, the Server MUST have full access right on all the Objects and Object Instances in the LwM2M Client, and the Access Control Object MAY be not instantiated."

It seems that we can just not instantiate the Access Control Object.

On the other side, I was curious about how these bits affect every instance. I found that even with all access bits in 0, I was able to perform all the actions. **Then I found [this](https://github.com/eclipse/wakaama/blob/31d64c0c41fae9653c1fa53ef58d1a44e49017fa/core/management.c#L196)**. It seems that Wakaama is actually ignoring the Access Control List.

Should we drop the instantiation of this object?

Reply:
Good finding! Yes, I would drop it for now. It's optional and Wakaama doesn't use it.